gdal/port/cpl_userfaultfd.cpp

```diff
@@ -0,0 +1,280 @@
+
+struct cpl_uffd_context {
+  bool keep_going = true;
+  int uffd = -1;
+  struct uffdio_register uffdio_register = {};
+  std::string filename = std::string("");
+  off_t  file_size = 0;
+  off_t  page_size = 0;
+  void * page_ptr = nullptr;
+  size_t vma_size = 0;
+  void * vma_ptr = nullptr;
+};
+
+
+static void uffd_cleanup(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+
+  if (!ctx) return;
+
+  if (ctx->uffd != -1) {
+    ioctl(ctx->uffd, UFFDIO_UNREGISTER, &ctx->uffdio_register);
+    close(ctx->uffd);
+  }
+  if (ctx->page_ptr && ctx->page_size)
+    munmap(ctx->page_ptr, ctx->page_size);
+  if (ctx->vma_ptr && ctx->vma_size)
+    munmap(ctx->vma_ptr, ctx->vma_size);
+
+  delete ctx;
+
+  return;
+}
+
+static int64_t get_page_limit()
+{
+  int64_t retval;
+  const char * variable = CPLGetConfigOption(GDAL_UFFD_LIMIT, nullptr);
+
+  if (variable && sscanf(variable, "%" PRId64, &retval))
+    return retval;
+  else
+    return -1;
+}
+
+static void * fault_handler(void * ptr)
+{
+  int64_t page_limit;
+  int64_t pages_used;
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+  struct uffd_msg msg;
+  struct uffdio_copy uffdio_copy;
+  struct pollfd pollfd;
+
+  // Machinery to limit the number of physical + swap pages consumed
+  page_limit = get_page_limit();
+  pages_used = 0;
+
+  // Register cleanup handler
+  pthread_cleanup_push(uffd_cleanup, ctx);
+
+  pollfd.fd = ctx->uffd;
+  pollfd.events = POLLIN;
+
+  // Open asset for reading
+  VSILFILE * file = VSIFOpenL(ctx->filename.c_str(), "rb");
+
+  if (!file) return nullptr;
+
+  // Loop until told to stop
+  while(ctx->keep_going) {
+    uintptr_t fault_addr;
+    uint64_t offset;
+    off_t bytes_needed;
+
+    // Poll for event
+    if (poll(&pollfd, 1, 16) == -1) break; // 60Hz when no demand
+    if ((pollfd.revents & POLLERR) || (pollfd.revents & POLLNVAL)) break;
+    if (!(pollfd.revents & POLLIN)) continue;
+
+    // Read page fault event
+    if (read(ctx->uffd, &msg, sizeof(msg)) < 1) {
+      if (errno == EWOULDBLOCK) continue;
+      else break;
+    }
+
+    fault_addr = msg.arg.pagefault.address & ~(ctx->page_size-1);
+    offset = static_cast<uint64_t>(fault_addr) - reinterpret_cast<uint64_t>(ctx->vma_ptr);
+    bytes_needed = static_cast<off_t>(ctx->file_size - offset);
+    if (bytes_needed > ctx->page_size) bytes_needed = ctx->page_size;
+
+    // Limit number of pages
+    if ((page_limit > 0) && (pages_used >= page_limit)) {
+      pages_used = 0;
+      ctx->page_ptr = mmap(ctx->page_ptr, ctx->page_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
+      if (ctx->page_ptr == BAD_MMAP) {
+        ctx->page_ptr = nullptr;
+        break;
+      }
+    }
+
+    // Copy data into page
+    int retries;
+    if (VSIFSeekL(file, offset, SEEK_SET)) break;
+    for (retries = 0; retries < 1000; ++retries) { // Try for up to 10 seconds
```

Comment:
Well, it is possible that I have made a mistake or remembered incorrectly, but I do recall seeing calls to `VSIFReadL` which were expected to return 1 (because one page in the middle of the file was being requested) returning 0.  The errors from HDF5 related to corrupt files seemed to go away when I made changes relative to that.

I'll double-check this evening.

Reply:
that could perhaps happen in case of network failure, but in that case the retry should be done by the /vsicurl/ level (there are configuration options for this)

Reply:
My recollection is that I observed the same behavior from the `/vsitar` driver with the same `.h5` file in a tarball.  Again, I could be mistaken so I will double-check in the evening.

Reply:
Okay, you were absolutely right: under normal circumstances, `VSIFReadL` always returns a record when one is available.

`VSIFReadL` can be observed to sometimes return zero records instead of one when the previous approach to `CPLDeleteUserFaultMapping` (just closing the uffd) is used.  It looks like that approach was just leading to memory corruption which was effecting `VSIFReadL` prior to it causing an outright segmentation fault.